<a href="https://colab.research.google.com/github/womaro/Story/blob/main/Seaborn_and_Linear_Regression_(start).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

Do higher film budgets lead to more box office revenue? Let's find out if there's a relationship using the movie budgets and financial performance data that I've scraped from [the-numbers.com](https://www.the-numbers.com/movie/budgets) on **May 1st, 2018**. 

<img src=https://i.imgur.com/kq7hrEh.png>

# Import Statements

In [1]:
import pandas as pd
import matplotlib.pyplot as plt


# Notebook Presentation

In [2]:
pd.options.display.float_format = '{:,.2f}'.format

from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()

# Read the Data

In [3]:
data = pd.read_csv('cost_revenue_dirty.csv')

# Explore and Clean the Data

**Challenge**: Answer these questions about the dataset:
1. How many rows and columns does the dataset contain?
2. Are there any NaN values present?
3. Are there any duplicate rows?
4. What are the data types of the columns?

In [4]:
data.shape

(5391, 6)

In [12]:
data.sample(10).sort_values('USD_Production_Budget', ascending=False)

Rank Release_Date                   Movie_Title USD_Production_Budget  \
1894  3756   10/18/2002                    Auto Focus            $7,000,000   
1269   731    7/16/1999                Eyes Wide Shut           $65,000,000   
2010   996    4/16/2003              Bulletproof Monk           $52,000,000   
1857  1046     8/5/2002                    Unfaithful           $50,000,000   
5300  1452   12/23/2016                      Why Him?           $38,000,000   
992   4266    8/13/1997                The Full Monty            $3,500,000   
1479  4336    8/18/2000  The Original Kings of Comedy            $3,000,000   
3177   149    7/11/2008   Madagascar: Escape 2 Africa          $150,000,000   
1018  5372    11/4/1997                Pink Flamingos               $12,000   
4218  4816    8/28/2012                    Below Zero            $1,200,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
1894          $2,703,821         $2,062,066  
1269        $104,267,443        $55,691,208  
2010         $23,010,607        $23,010,607  
1857        $119,114,494        $52,752,475  
5300        $117,523,786        $60,323,786  
992         $257,938,649        $45,950,122  
1479         $38,236,338        $38,168,022  
3177        $599,516,844       $180,010,950  
1018            $413,802           $413,802  
4218                  $0                 $0

In [13]:
data.isna().sum()

Rank                     0
Release_Date             0
Movie_Title              0
USD_Production_Budget    0
USD_Worldwide_Gross      0
USD_Domestic_Gross       0
dtype: int64

In [15]:
duplicates = data[data.duplicated()]
duplicates.head()

Empty DataFrame
Columns: [Rank, Release_Date, Movie_Title, USD_Production_Budget, USD_Worldwide_Gross, USD_Domestic_Gross]
Index: []

In [16]:
duplicates = data[data.duplicated(subset=['Movie_Title'])]
duplicates.head()

Rank Release_Date                   Movie_Title USD_Production_Budget  \
64   3973   12/23/1954  20,000 Leagues Under the Sea            $5,000,000   
84   2802   11/18/1959                       Ben-Hur           $15,000,000   
215  2202   12/17/1976                     King Kong           $23,000,000   
322  3105     2/4/1982                    Cat People           $12,500,000   
579  4154   10/19/1990      Night of the Living Dead            $4,200,000   

    USD_Worldwide_Gross USD_Domestic_Gross  
64          $28,200,000        $28,200,000  
84          $73,000,000        $73,000,000  
215         $90,614,445        $52,614,445  
322         $21,000,000         $7,000,000  
579          $5,835,247         $5,835,247

In [17]:
data[data.Movie_Title=='Ben-Hur']

Rank Release_Date Movie_Title USD_Production_Budget USD_Worldwide_Gross  \
5     4250   12/30/1925     Ben-Hur            $3,900,000          $9,000,000   
84    2802   11/18/1959     Ben-Hur           $15,000,000         $73,000,000   
5231   380    8/19/2016     Ben-Hur           $95,000,000         $88,725,100   

     USD_Domestic_Gross  
5            $9,000,000  
84          $73,000,000  
5231        $26,410,477

In [18]:
duplicates = data[data.duplicated(subset=['Movie_Title', 'Release_Date'])]
duplicates.head()

Rank Release_Date        Movie_Title USD_Production_Budget  \
3055  5290   12/31/2007  A Dog's Breakfast              $120,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
3055                  $0                 $0

In [19]:
data[data.Movie_Title=="A Dog's Breakfast"]

Rank Release_Date        Movie_Title USD_Production_Budget  \
3054  5289   12/31/2007  A Dog's Breakfast              $120,000   
3055  5290   12/31/2007  A Dog's Breakfast              $120,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
3054                  $0                 $0  
3055                  $0                 $0

In [20]:
data = data.drop_duplicates(subset=['Movie_Title', 'Release_Date'])

In [21]:
duplicates = data[data.duplicated(subset=['Movie_Title', 'Release_Date'])]
duplicates.head()

Empty DataFrame
Columns: [Rank, Release_Date, Movie_Title, USD_Production_Budget, USD_Worldwide_Gross, USD_Domestic_Gross]
Index: []

In [22]:
data.describe

<bound method NDFrame.describe of       Rank Release_Date                     Movie_Title USD_Production_Budget  \
0     5293     8/2/1915           The Birth of a Nation              $110,000   
1     5140     5/9/1916                     Intolerance              $385,907   
2     5230   12/24/1916    20,000 Leagues Under the Sea              $200,000   
3     5299    9/17/1920  Over the Hill to the Poorhouse              $100,000   
4     5222     1/1/1925                  The Big Parade              $245,000   
...    ...          ...                             ...                   ...   
5386  2950    10/8/2018                             Meg           $15,000,000   
5387   126   12/18/2018                         Aquaman          $160,000,000   
5388    96   12/31/2020                     Singularity          $175,000,000   
5389  1119   12/31/2020          Hannibal the Conqueror           $50,000,000   
5390  2517   12/31/2020  Story of Bonnie and Clyde, The           $20,000,0

### Data Type Conversions

**Challenge**: Convert the `USD_Production_Budget`, `USD_Worldwide_Gross`, and `USD_Domestic_Gross` columns to a numeric format by removing `$` signs and `,`. 
<br>
<br>
Note that *domestic* in this context refers to the United States.

**Challenge**: Convert the `Release_Date` column to a Pandas Datetime type. 

### Descriptive Statistics

**Challenge**: 

1. What is the average production budget of the films in the data set?
2. What is the average worldwide gross revenue of films?
3. What were the minimums for worldwide and domestic revenue?
4. Are the bottom 25% of films actually profitable or do they lose money?
5. What are the highest production budget and highest worldwide gross revenue of any film?
6. How much revenue did the lowest and highest budget films make?

# Investigating the Zero Revenue Films

**Challenge** How many films grossed $0 domestically (i.e., in the United States)? What were the highest budget films that grossed nothing?

**Challenge**: How many films grossed $0 worldwide? What are the highest budget films that had no revenue internationally?

### Filtering on Multiple Conditions

**Challenge**: Use the [`.query()` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html) to accomplish the same thing. Create a subset for international releases that had some worldwide gross revenue, but made zero revenue in the United States. 

Hint: This time you'll have to use the `and` keyword.

### Unreleased Films

**Challenge**:
* Identify which films were not released yet as of the time of data collection (May 1st, 2018).
* How many films are included in the dataset that have not yet had a chance to be screened in the box office? 
* Create another DataFrame called data_clean that does not include these films. 

In [ ]:
# Date of Data Collection
scrape_date = pd.Timestamp('2018-5-1')

### Films that Lost Money

**Challenge**: 
What is the percentage of films where the production costs exceeded the worldwide gross revenue? 

# Seaborn for Data Viz: Bubble Charts

### Plotting Movie Releases over Time

**Challenge**: Try to create the following Bubble Chart:

<img src=https://i.imgur.com/8fUn9T6.png>



# Converting Years to Decades Trick

**Challenge**: Create a column in `data_clean` that has the decade of the release. 

<img src=https://i.imgur.com/0VEfagw.png width=650> 

Here's how: 
1. Create a [`DatetimeIndex` object](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.html) from the Release_Date column. 
2. Grab all the years from the `DatetimeIndex` object using the `.year` property.
<img src=https://i.imgur.com/5m06Ach.png width=650>
3. Use floor division `//` to convert the year data to the decades of the films.
4. Add the decades as a `Decade` column to the `data_clean` DataFrame.

### Separate the "old" (before 1969) and "New" (1970s onwards) Films

**Challenge**: Create two new DataFrames: `old_films` and `new_films`
* `old_films` should include all the films before 1969 (up to and including 1969)
* `new_films` should include all the films from 1970 onwards
* How many films were released prior to 1970?
* What was the most expensive film made prior to 1970?

# Seaborn Regression Plots

**Challenge**: Use Seaborn's `.regplot()` to show the scatter plot and linear regression line against the `new_films`. 
<br>
<br>
Style the chart

* Put the chart on a `'darkgrid'`.
* Set limits on the axes so that they don't show negative values.
* Label the axes on the plot "Revenue in \$ billions" and "Budget in \$ millions".
* Provide HEX colour codes for the plot and the regression line. Make the dots dark blue (#2f4b7c) and the line orange (#ff7c43).

Interpret the chart

* Do our data points for the new films align better or worse with the linear regression than for our older films?
* Roughly how much would a film with a budget of $150 million make according to the regression line?

# Run Your Own Regression with scikit-learn

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**: Run a linear regression for the `old_films`. Calculate the intercept, slope and r-squared. How much of the variance in movie revenue does the linear model explain in this case?

# Use Your Model to Make a Prediction

We just estimated the slope and intercept! Remember that our Linear Model has the following form:

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**:  How much global revenue does our model estimate for a film with a budget of $350 million? 